File: Interviews/LeetCode/Topic-Wise/2-Pointers/Medium/3Sum-Closest/16.py
class Solution:
    def threeSumClosest(self, nums: List[int], target: int) -> int:
        ans, diff, n = -1, 1e9, len(nums)
        nums.sort()
        for i in range(n - 2):
            j, k = i + 1, n - 1
            while j < k:
                x = nums[i] + nums[j] + nums[k]
                if x == target: return x
                if abs(x - target) < diff:
                    diff = abs(x - target)
                    ans = x
                if x < target: j += 1
                else: k -= 1
        
        return ans

File: Interviews/LeetCode/Topic-Wise/2-Pointers/Medium/3Sum-Closest/16.cpp
class Solution {
public:
    int threeSumClosest(vector<int>& nums, int target) {
        int ans = -1, diff = 1e9, n = nums.size();
        sort(nums.begin(), nums.end());
        for (int i = 0; i < n - 2; i++) {
            int j = i + 1, k = n - 1;
            while (j < k) {
                int x = nums[i] + nums[j] + nums[k];
                if (x == target) {
                    return x;
                }
                if (abs(x - target) < diff) {
                    diff = abs(x - target);
                    ans = x;
                }

                if (x < target) {
                    j += 1;
                } else {
                    k -= 1;
                }
            }
        }

        return ans;
    }
};